File: fontes/frontend/fenix_provider/lib/src/model/tributacao/tribut_icms_uf.dart
/*
Title: T2Ti ERP 3.0                                                                
Description: Model relacionado à tabela [TRIBUT_ICMS_UF] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2021 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************/
import 'dart:convert';


class TributIcmsUf {
	int? id;
	int? idTributConfiguraOfGt;
	String? ufDestino;
	int? cfop;
	String? csosn;
	String? cst;
	String? modalidadeBc;
	double? aliquota;
	double? valorPauta;
	double? valorPrecoMaximo;
	double? mva;
	double? porcentoBc;
	String? modalidadeBcSt;
	double? aliquotaInternaSt;
	double? aliquotaInterestadualSt;
	double? porcentoBcSt;
	double? aliquotaIcmsSt;
	double? valorPautaSt;
	double? valorPrecoMaximoSt;

	TributIcmsUf({
		id,
		idTributConfiguraOfGt,
		ufDestino,
		cfop,
		csosn,
		cst,
		modalidadeBc,
		aliquota = 0.0,
		valorPauta = 0.0,
		valorPrecoMaximo = 0.0,
		mva = 0.0,
		porcentoBc = 0.0,
		modalidadeBcSt,
		aliquotaInternaSt = 0.0,
		aliquotaInterestadualSt = 0.0,
		porcentoBcSt = 0.0,
		aliquotaIcmsSt = 0.0,
		valorPautaSt = 0.0,
		valorPrecoMaximoSt = 0.0,
	});

	static List<String> campos = <String>[
		'ID', 
		'UF_DESTINO', 
		'CFOP', 
		'CSOSN', 
		'CST', 
		'MODALIDADE_BC', 
		'ALIQUOTA', 
		'VALOR_PAUTA', 
		'VALOR_PRECO_MAXIMO', 
		'MVA', 
		'PORCENTO_BC', 
		'MODALIDADE_BC_ST', 
		'ALIQUOTA_INTERNA_ST', 
		'ALIQUOTA_INTERESTADUAL_ST', 
		'PORCENTO_BC_ST', 
		'ALIQUOTA_ICMS_ST', 
		'VALOR_PAUTA_ST', 
		'VALOR_PRECO_MAXIMO_ST', 
	];
	
	static List<String> colunas = <String>[
		'Id', 
		'UF', 
		'CFOP', 
		'CSOSN', 
		'CST', 
		'Modalidade Base Cálculo', 
		'Alíquota', 
		'Valor Pauta', 
		'Valor Preço Máximo', 
		'Valor MVA', 
		'Porcento Base Cálculo', 
		'Modalidade Base Cálculo ST', 
		'Alíquota Interna ST', 
		'Alíquota Interestadual ST', 
		'Porcento Base Cálculo ST', 
		'Alíquota ICMS ST', 
		'Valor Pauta ST', 
		'Valor Preço Máximo ST', 
	];

	TributIcmsUf.fromJson(Map<String, dynamic> jsonDados) {
		id = jsonDados['id'];
		idTributConfiguraOfGt = jsonDados['idTributConfiguraOfGt'];
		ufDestino = jsonDados['ufDestino'] == '' ? null : jsonDados['ufDestino'];
		cfop = jsonDados['cfop'];
		csosn = jsonDados['csosn'];
		cst = jsonDados['cst'];
		modalidadeBc = getModalidadeBc(jsonDados['modalidadeBc']);
		aliquota = jsonDados['aliquota']?.toDouble();
		valorPauta = jsonDados['valorPauta']?.toDouble();
		valorPrecoMaximo = jsonDados['valorPrecoMaximo']?.toDouble();
		mva = jsonDados['mva']?.toDouble();
		porcentoBc = jsonDados['porcentoBc']?.toDouble();
		modalidadeBcSt = getModalidadeBcSt(jsonDados['modalidadeBcSt']);
		aliquotaInternaSt = jsonDados['aliquotaInternaSt']?.toDouble();
		aliquotaInterestadualSt = jsonDados['aliquotaInterestadualSt']?.toDouble();
		porcentoBcSt = jsonDados['porcentoBcSt']?.toDouble();
		aliquotaIcmsSt = jsonDados['aliquotaIcmsSt']?.toDouble();
		valorPautaSt = jsonDados['valorPautaSt']?.toDouble();
		valorPrecoMaximoSt = jsonDados['valorPrecoMaximoSt']?.toDouble();
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonDados = <String, dynamic>{};

		jsonDados['id'] = id ?? 0;
		jsonDados['idTributConfiguraOfGt'] = idTributConfiguraOfGt ?? 0;
		jsonDados['ufDestino'] = ufDestino;
		jsonDados['cfop'] = cfop ?? 0;
		jsonDados['csosn'] = csosn;
		jsonDados['cst'] = cst;
		jsonDados['modalidadeBc'] = setModalidadeBc(modalidadeBc);
		jsonDados['aliquota'] = aliquota;
		jsonDados['valorPauta'] = valorPauta;
		jsonDados['valorPrecoMaximo'] = valorPrecoMaximo;
		jsonDados['mva'] = mva;
		jsonDados['porcentoBc'] = porcentoBc;
		jsonDados['modalidadeBcSt'] = setModalidadeBcSt(modalidadeBcSt);
		jsonDados['aliquotaInternaSt'] = aliquotaInternaSt;
		jsonDados['aliquotaInterestadualSt'] = aliquotaInterestadualSt;
		jsonDados['porcentoBcSt'] = porcentoBcSt;
		jsonDados['aliquotaIcmsSt'] = aliquotaIcmsSt;
		jsonDados['valorPautaSt'] = valorPautaSt;
		jsonDados['valorPrecoMaximoSt'] = valorPrecoMaximoSt;
	
		return jsonDados;
	}
	
    getModalidadeBc(String? modalidadeBc) {
    	switch (modalidadeBc) {
    		case '0':
    			return '0-Margem Valor Agregado (%)';
    		case '1':
    			return '1-Pauta (Valor)';
    		case '2':
    			return '2-Preço Tabelado Máx. (valor)';
    		case '3':
    			return '3-Valor da Operação';
    		default:
    			return null;
    		}
    	}

    setModalidadeBc(String? modalidadeBc) {
      if (modalidadeBc?.length == 1) {
        return modalidadeBc;
      }
    	switch (modalidadeBc) {
    		case '0-Margem Valor Agregado (%)':
    			return '0';
    		case '1-Pauta (Valor)':
    			return '1';
    		case '2-Preço Tabelado Máx. (valor)':
    			return '2';
    		case '3-Valor da Operação':
    			return '3';
    		default:
    			return null;
    		}
    	}

    getModalidadeBcSt(String? modalidadeBcSt) {
    	switch (modalidadeBcSt) {
    		case '0':
    			return '0-Preço tabelado ou máximo sugerido';
    		case '1':
    			return '1-Lista Negativa (valor)';
    		case '2':
    			return '2-Lista Positiva (valor)';
    		case '3':
    			return '3-Lista Neutra (valor)';
    		case '4':
    			return '4-Margem Valor Agregado (%)';
    		case '5':
    			return '5-Pauta (valor)';
    		default:
    			return null;
    		}
    	}

    setModalidadeBcSt(String? modalidadeBcSt) {
      if (modalidadeBcSt?.length == 1) {
        return modalidadeBcSt;
      }
      switch (modalidadeBcSt) {
    		case '0-Preço tabelado ou máximo sugerido':
    			return '0';
    		case '1-Lista Negativa (valor)':
    			return '1';
    		case '2-Lista Positiva (valor)':
    			return '2';
    		case '3-Lista Neutra (valor)':
    			return '3';
    		case '4-Margem Valor Agregado (%)':
    			return '4';
    		case '5-Pauta (valor)':
    			return '5';
    		default:
    			return null;
    		}
    	}


	String objetoEncodeJson(TributIcmsUf objeto) {
	  final jsonDados = objeto.toJson;
	  return json.encode(jsonDados);
	}
	
}

File: fontes/backend/php/fenix/src/model/tributacao/TributIcmsUf.php
<?php
/*******************************************************************************
Title: T2Ti ERP 3.0
Description: Model relacionado à tabela [TRIBUT_ICMS_UF] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2021 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************/
declare(strict_types=1);

class TributIcmsUf extends EloquentModel implements \JsonSerializable
{
    /**
     * The table associated with the model.
     *
     * @var string
     */
    protected $table = 'TRIBUT_ICMS_UF';

    /**
     * Eager Loading - Relacionamentos que devem ser sempre carregados por padrão
     *
     * @var array
     */
	
    /**
     * Relations
     */
    public function tributConfiguraOfGt()
    {
    	return $this->belongsTo(TributConfiguraOfGt::class, 'ID_TRIBUT_CONFIGURA_OF_GT', 'ID');
    }


    /**
     * Gets e Sets
     */
    public function getIdAttribute() 
	{
		return $this->attributes['ID'];
	}

	public function setIdAttribute($id) 
	{
		$this->attributes['ID'] = $id;
	}

    public function getUfDestinoAttribute() 
	{
		return $this->attributes['UF_DESTINO'];
	}

	public function setUfDestinoAttribute($ufDestino) 
	{
		$this->attributes['UF_DESTINO'] = $ufDestino;
	}

    public function getCfopAttribute() 
	{
		return $this->attributes['CFOP'];
	}

	public function setCfopAttribute($cfop) 
	{
		$this->attributes['CFOP'] = $cfop;
	}

    public function getCsosnAttribute() 
	{
		return $this->attributes['CSOSN'];
	}

	public function setCsosnAttribute($csosn) 
	{
		$this->attributes['CSOSN'] = $csosn;
	}

    public function getCstAttribute() 
	{
		return $this->attributes['CST'];
	}

	public function setCstAttribute($cst) 
	{
		$this->attributes['CST'] = $cst;
	}

    public function getModalidadeBcAttribute() 
	{
		return $this->attributes['MODALIDADE_BC'];
	}

	public function setModalidadeBcAttribute($modalidadeBc) 
	{
		$this->attributes['MODALIDADE_BC'] = $modalidadeBc;
	}

    public function getAliquotaAttribute() 
	{
		return (double)$this->attributes['ALIQUOTA'];
	}

	public function setAliquotaAttribute($aliquota) 
	{
		$this->attributes['ALIQUOTA'] = $aliquota;
	}

    public function getValorPautaAttribute() 
	{
		return (double)$this->attributes['VALOR_PAUTA'];
	}

	public function setValorPautaAttribute($valorPauta) 
	{
		$this->attributes['VALOR_PAUTA'] = $valorPauta;
	}

    public function getValorPrecoMaximoAttribute() 
	{
		return (double)$this->attributes['VALOR_PRECO_MAXIMO'];
	}

	public function setValorPrecoMaximoAttribute($valorPrecoMaximo) 
	{
		$this->attributes['VALOR_PRECO_MAXIMO'] = $valorPrecoMaximo;
	}

    public function getMvaAttribute() 
	{
		return (double)$this->attributes['MVA'];
	}

	public function setMvaAttribute($mva) 
	{
		$this->attributes['MVA'] = $mva;
	}

    public function getPorcentoBcAttribute() 
	{
		return (double)$this->attributes['PORCENTO_BC'];
	}

	public function setPorcentoBcAttribute($porcentoBc) 
	{
		$this->attributes['PORCENTO_BC'] = $porcentoBc;
	}

    public function getModalidadeBcStAttribute() 
	{
		return $this->attributes['MODALIDADE_BC_ST'];
	}

	public function setModalidadeBcStAttribute($modalidadeBcSt) 
	{
		$this->attributes['MODALIDADE_BC_ST'] = $modalidadeBcSt;
	}

    public function getAliquotaInternaStAttribute() 
	{
		return (double)$this->attributes['ALIQUOTA_INTERNA_ST'];
	}

	public function setAliquotaInternaStAttribute($aliquotaInternaSt) 
	{
		$this->attributes['ALIQUOTA_INTERNA_ST'] = $aliquotaInternaSt;
	}

    public function getAliquotaInterestadualStAttribute() 
	{
		return (double)$this->attributes['ALIQUOTA_INTERESTADUAL_ST'];
	}

	public function setAliquotaInterestadualStAttribute($aliquotaInterestadualSt) 
	{
		$this->attributes['ALIQUOTA_INTERESTADUAL_ST'] = $aliquotaInterestadualSt;
	}

    public function getPorcentoBcStAttribute() 
	{
		return (double)$this->attributes['PORCENTO_BC_ST'];
	}

	public function setPorcentoBcStAttribute($porcentoBcSt) 
	{
		$this->attributes['PORCENTO_BC_ST'] = $porcentoBcSt;
	}

    public function getAliquotaIcmsStAttribute() 
	{
		return (double)$this->attributes['ALIQUOTA_ICMS_ST'];
	}

	public function setAliquotaIcmsStAttribute($aliquotaIcmsSt) 
	{
		$this->attributes['ALIQUOTA_ICMS_ST'] = $aliquotaIcmsSt;
	}

    public function getValorPautaStAttribute() 
	{
		return (double)$this->attributes['VALOR_PAUTA_ST'];
	}

	public function setValorPautaStAttribute($valorPautaSt) 
	{
		$this->attributes['VALOR_PAUTA_ST'] = $valorPautaSt;
	}

    public function getValorPrecoMaximoStAttribute() 
	{
		return (double)$this->attributes['VALOR_PRECO_MAXIMO_ST'];
	}

	public function setValorPrecoMaximoStAttribute($valorPrecoMaximoSt) 
	{
		$this->attributes['VALOR_PRECO_MAXIMO_ST'] = $valorPrecoMaximoSt;
	}


    /**
     * Mapping
     */
    public function mapear($objeto)
    {
        if (isset($objeto)) {
            isset($objeto->id) ? $this->setIdAttribute($objeto->id) : $this->setIdAttribute(null);

			$this->setUfDestinoAttribute($objeto->ufDestino);
			$this->setCfopAttribute($objeto->cfop);
			$this->setCsosnAttribute($objeto->csosn);
			$this->setCstAttribute($objeto->cst);
			$this->setModalidadeBcAttribute($objeto->modalidadeBc);
			$this->setAliquotaAttribute($objeto->aliquota);
			$this->setValorPautaAttribute($objeto->valorPauta);
			$this->setValorPrecoMaximoAttribute($objeto->valorPrecoMaximo);
			$this->setMvaAttribute($objeto->mva);
			$this->setPorcentoBcAttribute($objeto->porcentoBc);
			$this->setModalidadeBcStAttribute($objeto->modalidadeBcSt);
			$this->setAliquotaInternaStAttribute($objeto->aliquotaInternaSt);
			$this->setAliquotaInterestadualStAttribute($objeto->aliquotaInterestadualSt);
			$this->setPorcentoBcStAttribute($objeto->porcentoBcSt);
			$this->setAliquotaIcmsStAttribute($objeto->aliquotaIcmsSt);
			$this->setValorPautaStAttribute($objeto->valorPautaSt);
			$this->setValorPrecoMaximoStAttribute($objeto->valorPrecoMaximoSt);

			// vincular objetos
        }
    }

    /**
     * Serialization
     * {@inheritdoc}
     */
    public function jsonSerialize()
    {
        return [
			'id' => $this->getIdAttribute(),
			'ufDestino' => $this->getUfDestinoAttribute(),
			'cfop' => $this->getCfopAttribute(),
			'csosn' => $this->getCsosnAttribute(),
			'cst' => $this->getCstAttribute(),
			'modalidadeBc' => $this->getModalidadeBcAttribute(),
			'aliquota' => $this->getAliquotaAttribute(),
			'valorPauta' => $this->getValorPautaAttribute(),
			'valorPrecoMaximo' => $this->getValorPrecoMaximoAttribute(),
			'mva' => $this->getMvaAttribute(),
			'porcentoBc' => $this->getPorcentoBcAttribute(),
			'modalidadeBcSt' => $this->getModalidadeBcStAttribute(),
			'aliquotaInternaSt' => $this->getAliquotaInternaStAttribute(),
			'aliquotaInterestadualSt' => $this->getAliquotaInterestadualStAttribute(),
			'porcentoBcSt' => $this->getPorcentoBcStAttribute(),
			'aliquotaIcmsSt' => $this->getAliquotaIcmsStAttribute(),
			'valorPautaSt' => $this->getValorPautaStAttribute(),
			'valorPrecoMaximoSt' => $this->getValorPrecoMaximoStAttribute(),
        ];
    }
}